### PR TITLE
#158757927 Create pre-commit hook

### PR DESCRIPTION
**Pull request by ibutiti**
#### What does this PR do?
- Migrate the project from managing dependencies in a `requirements.txt` file to using `pipfiles`. This separates development and production dependencies in an elegant and easy to manage way.
- Create a pre-commit hook to auto format python code to `pep8` using `yapf` then check for residual linting errors using `pylint`.

#### Description of Task to be completed?
- Create way to easily manage development and production dependencies. The linting and formatting dependencies should reside in the development dependencies separate from production dependencies.

- Create pre commit hook that automatically formats the code to `pep8` and highlights linting errors not automatically resolved.

#### How should this be manually tested?
- Create a local copy with the following commands: `git fetch origin ch-create-pre-commit-hook-158757927:ch-create-pre-commit-hook-158757927`, then `git checkout ch-create-pre-commit-hook-158757927`
- Follow the instructions on how to setup the pre-commit hook into your own repository. These instructions are in the `Contribution Guide` section in the Readme file.
- Once your pre-commit hook has been setup, create a test branch and make changes contrary to `pep8` to any `.py` file and attempt to commit it without staging it. You should get an error stating `You have unstaged changes to some files in your commit;`
- Add the changed files to staging and attempt to make a commit. The pre-commit hook should be called and auto-format the `.py` files to `pep8` then call `pylint` on the files. If there are any linting issues, the commit will be blocked until they are resolved.

#### Any background context you want to provide?
- N/A

#### What are the relevant pivotal tracker stories?
[#158757927](https://www.pivotaltracker.com/story/show/158757927)

#### Screenshots (if appropriate)
- N/A

#### Questions:
- N/A

### CODE REVIEW

File: .style.yapf [hunk 1]
```diff
@@ -0,0 +1,2 @@
+[style]
+based_on_style = pep8
```
Comment (SEUNAGBEYE) [inline .style.yapf]: @ibutiti add a new line please

File: Pipfile [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+flask-migrate = "*"
+coverage = "*"
+marshmallow = "*"
+numpy = "*"
+psycopg2 = "*"
+pyjwt = "*"
+pytest = "*"
+pytest-cov = "*"
+python-dotenv = "*"
+cryptography = "*"
```
Comment (SEUNAGBEYE) [inline Pipfile]: @ibutiti are these all the dependencies we currently have in the application?

File: Readme.md [hunk 3]
```diff
@@ -14,72 +14,89 @@ An asset-management tool for Andela
...
+
     ```
-- Create virtual env
-    ```
-    virtualenv --python=python3 venv
-    ```
-- Activate virtual env
+- Install dependencies:
     ```
-    source venv/bin/activate
```
Comment (SEUNAGBEYE) [inline Readme.md]: Hmm, goodbye virtualenvs

File: Readme.md [hunk 4]
```diff
@@ -14,72 +14,89 @@ An asset-management tool for Andela
...
+    pipenv install
     ```
-- Install dependencies
+
+- Install dev dependencies to setup development environment:
     ```
-    pip install -r requirements.txt
+    pipenv install --dev
```
Comment (SEUNAGBEYE) [inline Readme.md]: Nice

File: Pipfile (hunk 2 above)
Comment (ibutiti) [reply Pipfile]: These are the main dependencies. Their dependencies are in the `pipfile.lock` file.

File: Readme.md (hunk 3 above)
Comment (ibutiti) [reply Readme.md]: Good riddance 😄

File: Pipfile (hunk 2 above)
Comment (SEUNAGBEYE) [reply Pipfile]: @ibutiti make sense

File: .style.yapf (hunk 1 above)
Comment (ibutiti) [reply .style.yapf]: Done

**Comment by ibutiti** [comment]
@Victoradukwu Done. Sorry, forgot to paste the instructions.

File: hooks/pre-commit [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+fi
+
+# Format all staged files, then exit with an error code if any have uncommitted
+# changes.
+echo
+echo 'Formatting staged Python files . . .'
+echo
+
+########## PIP VERSION #############
+# yapf -i -r -vv ${PYTHON_FILES[@]}
```
Comment (brayoh) [inline hooks/pre-commit]: @ibutiti kindly remove this if it's not been used

File: hooks/pre-commit [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+fi
+
+########## PIP VERSION #############
+# Verify that yapf is installed; if not, warn and exit.
+# if [ -z $(which yapf) ]; then
+#   echo 'yapf not on path; can not format. Please install yapf:'
+#   echo '    pip install yapf'
+#   exit 2
+# fi
+######### END PIP VERSION ##########
```
Comment (brayoh) [inline hooks/pre-commit]: and this too

File: main.py [hunk 7]
```diff
@@ -48,4 +48,4 @@ def handle_exception(error):
 
     response = jsonify(error.to_dict())
     response.status_code = error.status_code
-    return response
\ No newline at end of file
+    return response
```
Comment (brayoh) [inline main.py]: add new line here at the end of the file

File: hooks/pre-commit (hunk 5 above)
Comment (ibutiti) [reply hooks/pre-commit]: On it

File: hooks/pre-commit (hunk 6 above)
Comment (ibutiti) [reply hooks/pre-commit]: On it

File: main.py (hunk 7 above)
Comment (ibutiti) [reply main.py]: There's an end of file line here. I think it wasn't there before

**Comment by ibutiti** [comment]
@brayoh @SEUNAGBEYE requested changes made, and `.editorconfig` file added

File: .editorconfig [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+indent_style = space
+indent_size = 4
+insert_final_newline = true
+trim_trailing_whitespace = true
+end_of_line = lf
+charset = utf-8
+
+# Docstrings and comments use max_line_length = 79
+[*.py]
+max_line_length = 119
```
Comment (Dheavyman) [inline .editorconfig]: Just curious, why using 119 as the max line length. Does this not go against the default max of 79 for a line that we are adhering to?

Comment (brayoh) [reply .editorconfig]: this should be set to 80

File: .editorconfig [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+# Makefiles always use tabs for indentation
+[Makefile]
+indent_style = tab
+
+# Batch files use tabs for indentation
+[*.bat]
+indent_style = tab
+
+[docs/**.txt]
+max_line_length = 79
```
Comment (brayoh) [inline .editorconfig]: remove the uneeded configs, Makefile, js, json, txt, and html

File: .editorconfig (hunk 8 above)
Comment (ibutiti) [reply .editorconfig]: Cool @brayoh @Dheavyman

File: .editorconfig (hunk 9 above)
Comment (ibutiti) [reply .editorconfig]: cool @brayoh

**Comment by ibutiti** [comment]
Changes done @brayoh @Dheavyman

File: .editorconfig [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+trim_trailing_whitespace = true
+end_of_line = lf
+charset = utf-8
+
+# Docstrings and comments use max_line_length = 79
+[*.py]
+max_line_length = 79
+
+[docs/**.txt]
+max_line_length = 79
```
Comment (brayoh) [inline .editorconfig]: do we have any txt files in the codebase?

File: Readme.md [hunk 11]
```diff
@@ -90,5 +107,15 @@ When contributing to this repository, please first discuss the change you wish t
...
+- Contributors are required to activate the git pre-commit hook to auto format staged Python files to pep8 with yapf and check for residual pep8 linting errors using pylint.
+All commits are required to pass all checks from the pre-commit hook.
+The pre-commit hook can be installed as follows:
+Option 1: Copy the `hooks/pre-commit` file into the `.git/hooks` directory.
+          You will need to do this every time the `hooks/pre-commit` file is changed.
+Option 2: Create a file `.git/hooks/pre-commit` then create a symlink to this file by running the command:
+          `ln -s -f ../../hooks/pre-commit .git/hooks/pre-commit`
+          You will only need to do this once for your local repository.
+- Although highly discouraged, the pre-commit hook can be bypassed by passing the `--no-verify` flag to the commit command as follows:
+`git commit --no-verify -m "commit message"`
```
Comment (brayoh) [inline Readme.md]: Also is this included in the wiki?

File: .editorconfig (hunk 10 above)
Comment (ibutiti) [reply .editorconfig]: Yeah. Requirements.txt

Comment (brayoh) [reply .editorconfig]: Cool

File: Readme.md (hunk 11 above)
Comment (ibutiti) [reply Readme.md]: Included now @brayoh

Comment (brayoh) [reply Readme.md]: Awesome

**Pull request merged**